src/components/Alarm/Alarm.tsx

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Cognite AS
+import React, { FC, useRef, useEffect } from 'react';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemText from '@material-ui/core/ListItemText';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
+import MenuItem from '@material-ui/core/MenuItem';
+import Menu from '@material-ui/core/Menu';
+import IconButton from '@material-ui/core/IconButton';
+import Badge from '@material-ui/core/Badge';
+import CloseIcon from '@material-ui/icons/Close';
+import RadioButtonCheckedIcon from '@material-ui/icons/RadioButtonChecked';
+import './Alarm.css';
+import { RootState, RootAction } from 'StoreTypes';
+import { connect } from 'react-redux';
+import { Dispatch, bindActionCreators } from 'redux';
+import { AlarmType } from './interfaces';
+import {
+  startUpdateAlarms,
+  stopUpdateAlarms,
+  removeAlarm,
+} from '../../store/actions/root-action';
+
+/**
+ * This is the main functional component renders the alarm widget
+ */
+const Alarm: FC<Props> = (props: Props) => {
+  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+  const [selectedIndex, setSelectedIndex] = React.useState(0);
+  const alarmElement = useRef<HTMLDivElement>(null);
+
+  /**
+   * This will activate on alarms count badge click
+   * and display the alarm list in a dropdown menu
+   */
+  const onMoreAlarmsClick = () => {
+    setAnchorEl(alarmElement.current);
+  };
+
+  /**
+   * This will activate on alarm item click in the dropdown menu and
+   * set the clicked alarm as the selected alarm
+   */
+  const handleMenuItemClick = (index: number) => {
+    setSelectedIndex(index);
+    onMoreAlarmsClose();
+  };
+
+  /**
+   * This will activate on remove icon click in the alarm list.
+   * 'Remove Alarm' dipatch action will be fired and list's first alarm will be selected as the default alarm.
+   */
+  const onRemoveAlarmClick = (index: number, alarmId: number) => {
+    if (index !== 0 && index !== selectedIndex) {
+      setSelectedIndex(0);
+    }
+    props.removeAlarm(alarmId);
+  };
+
+  /**
+   * This is used to hide the alarm list dropdown menu
+   */
+  const onMoreAlarmsClose = () => {
+    setAnchorEl(null);
+  };
+
+  /**
+   * This is used to get the same width and height of selected alarm div
+   * to the alarm items in the dropdown list
+   */
+  const menuItemStyles = () => {
+    const alarmEl = alarmElement.current;
+    return alarmEl
+      ? {
+          width: `${alarmEl.clientWidth}px`,
+          height: `${alarmEl.clientHeight}px`,
+        }
+      : {};
+  };
+
+  /**
+   * Incase if component is umounted, alarms polling will be ended
+   */
+  const onUnmount = () => {
+    props.stopUpdateAlarms();
```

Comment:
In other widgets, it handles in a common place. But how the actions are fired is same.

Reply:
I really dont see why you say Alarm widget is different. How is it different? Only difference I see is it is not shown in the grid. Apart from that it is quite similar to others - It has a UI, it needs data fetched from back end. It would be better if you could stick to one design.

Reply:
Design is same. But we cannot fire these actions in same place where other widget fires. There it fires when the grid renders. But this renders in another place. Also other widgets are generates based on a list. But this isn't linked to that list

Reply:
I agree that actions cannot be fired in the same place. However for widgets firing actions was NOT a responsibility of the widget it self, which was good. Cant we move the firing actions outside of the UI component - to the start up sequence?

Reply:
When the alarm widget is unmounted, we need to fire an action to stop the alarms polling. So this is the only place we can use as a function which triggers when the component is unmounted.

Reply:
you are transferring the responsibility of data fetch control to the UI element which is not what we want. This widget gets unmounted when app is closed, or when the view changes, etc. Those are better places to handle the firing. Can you create a task so we can have a discussion around this later?